### ISSUE TITLE
Change jsonencode() to disable escaping of <, >, and &

### ISSUE THREAD

**Issue by dawidpogorzelski** [hashicorp/terraform#26110]
### Terraform Version

```
0.13.2
```

### Terraform Configuration Files

```terraform
> jsonencode({aa=">aaaa"})
{"aa":"\u003eaaaa"}
```

### Expected Behavior

the final result should have been `{"aa":">aaaa"}`

**Comment by alisdair** [comment]
Thanks for reporting this!

This is confusing looking output, but it is technically within the JSON spec. `\u003e` is a valid encoding of `>`. The source of this behaviour is [the Go `encoding/json` package](https://golang.org/src/encoding/json/encode.go#L46):

> String values encode as JSON strings coerced to valid UTF-8,
> replacing invalid bytes with the Unicode replacement rune.
> So that the JSON will be safe to embed inside HTML `<script>` tags,
> the string is encoded using HTMLEscape,
> which replaces "`<`", "`>`", "`&`", `U+2028`, and `U+2029` are escaped
> to "\u003c","\u003e", "\u0026", "\u2028", and "\u2029".
> This replacement can be disabled when using an Encoder,
> by calling SetEscapeHTML(false).

Because the Go team's rationale doesn't really apply in Terraform—hopefully you're not rendering outputs inside a `<script>` tag—it would seem reasonable to me to remove this escaping from the `jsonencode` function. If we were to choose to do this, I think it would be done in [the upstream go-cty package](https://github.com/zclconf/go-cty).

As a result I'm marking this issue as an enhancement, and retitling it for easier searching in future.

More details, along with a workaround using `replace`, are in this previous issue: #18871

**Comment by apparentlymart** [comment]
While I do see the human-readability advantages of avoiding escaping when it isn't necessary, I think we're strongly constrained here by the fact that `jsonencode` results are often used to populate resource arguments that can only be changed by destroying and re-creating the object they belong to, and therefore `jsonencode`'s results must remain character-for-character consistent for a given input in future versions, unless the need for a change is so severe as to outweigh the pain of forcing object replacement.

This compatibility constraint is the same thing that [the `yamlencode` function](https://www.terraform.io/docs/configuration/functions/yamlencode.html) is currently _avoiding_ being bound to by the warning in its docs (at the time I'm writing this comment) of its exact formatting being subject to change in future releases:

> **Warning:** This function is currently experimental and its exact result format may change in future versions of Terraform, based on feedback. Do not use `yamldecode` to construct a value for any resource argument where changes to the result would be disruptive. To get a consistent string representation of a value use `jsonencode` instead; its results are also valid YAML because YAML is a JSON superset.

Because the decision to either escape or not is a purely cosmetic decision that doesn't affect the meaning of the resulting document for a valid JSON parser, unfortunately I don't think it would be justified to force object replacement for existing objects in order to change that decision now.

Reviewing the comment I left in the other issue @alisdair mentioned (#18871) makes me suspect that the current behavior exists for the same reason I gave above: when we rewrite `jsonencode` for Terraform 0.12, we aimed to make it produce identical results to Terraform 0.11's `jsonencode` for any value that the old version (which was more limited) would've accepted. Terraform 0.11's `jsonencode` was just a thin wrapper around Go's `json.Marshal`, and so the new Terraform-specific `jsonencode` implementation in 0.12 is (I believe) intentionally following that behavior in order to preserve backward compatibility.
